File: app/src/main/java/com/originb/inkwisenote/common/ListUtils.java
package com.originb.inkwisenote.common;

import java.util.*;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ListUtils {
    public static <T> List<T> listOf(T... items) {
        if (items == null) return new ArrayList<>();
        switch (items.length) {
            case 0:
                return Collections.emptyList();
            case 1:
                return Collections.singletonList(items[0]);
            default:
                return Collections.unmodifiableList(Arrays.asList(items));
        }
    }

    public static <T> List<T> merge(List<T> a, List<T> b) {
        if (a == null && b != null) return b;
        else if (a != null && b == null) return a;
        else if (a == null && b == null) return new ArrayList<>();

        a.addAll(b);
        return a;
    }

    public static <T, K> Map<K, List<T>> groupBy(List<T> list, Function<? super T, ? extends K> keyMapper) {
        return list.stream().collect(Collectors.<T, K, List<T>>toMap(keyMapper, ListUtils::listOf, ListUtils::merge));
    }
}

File: shared/src/commonMain/kotlin/org/basnalcorp/shared/util/ListUtils.kt
package org.basnalcorp.shared.util

object ListUtils {
    fun <T> listOf(vararg items: T?): List<T?> =
        items.toList()

    fun <T> merge(a: List<T?>?, b: List<T?>?): List<T?> {
        val result = mutableListOf<T?>()
        a?.let { result.addAll(it) }
        b?.let { result.addAll(it) }
        return result
    }

    fun <T, K> groupBy(list: List<T>, keyMapper: (T) -> K): Map<K, List<T>> =
        list.groupBy(keyMapper)
}
